### PR TITLE
fix: [MR-523] Prevent duplicates of aborted or paused responses

### PR DESCRIPTION
**Pull request by alin-at-dfinity**
[MR-523]: While a response is in DTS execution, we no longer have a response with that `CallbackId` in `CanisterQueues` (meaning we wouldn't detect it as a duplicate callback); but we still have the callback itself (so that a second response would match an existing callback).

Explicitly check against the `CallbackId` of a potential aborted or paused response in `SystemState::push_input()` and `SystemState::induct_messages_to_self()`.

[MR-523]: https://dfinity.atlassian.net/browse/MR-523?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: rs/replicated_state/src/canister_state/system_state.rs [hunk 1]
```diff
@@ -1405,10 +1408,22 @@ impl SystemState {
...
                 && can_push(msg, *subnet_available_memory).is_err()
             {
-                // Bail out if not enough memory available for message.
                 return;
             }
 
+            // Protect against enqueuing a second response for the currently executing
+            // (aborted or paused) callback.
+            if let RequestOrResponse::Response(response) = &msg {
+                if let Some(aboirted_or_paused_response) = self.aborted_or_paused_response() {
```
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/system_state.rs:1417]:
```suggestion
                if let Some(aborted_or_paused_response) = self.aborted_or_paused_response() {
```

File: rs/replicated_state/src/canister_state/system_state.rs [hunk 2]
```diff
@@ -1405,10 +1408,22 @@ impl SystemState {
...
+                    if response.originator_reply_callback
+                        == aboirted_or_paused_response.originator_reply_callback
+                    {
+                        // The callback is already executing, don't enqueue a second response for it.
+                        return;
```
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/system_state.rs:1422]: Is returning here the right thing to do? I suppose returning means that things get handled via the loopback stream leading to eventually dropping the duplicate message upon induction?

Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/system_state.rs:1422]: Returning stops inducting messages. We peeked a message we didn't like and returned without popping it or the messages after it.

**Review by derlerd-dfinity — APPROVED** [review]
Assuming we still add appropriate tests this change looks good to me.

**Pull request merged**